packages/core/src/main/conversation/ConversationService.ts

```diff
@@ -1045,126 +1077,104 @@ export class ConversationService {
+      };
+    } catch (error) {
+      return {
+        ...payload,
+        content,
+        messageTimer: genericMessage.ephemeral?.expireAfterMillis || 0,
+        state: PayloadBundleState.CANCELLED,
+      };
     }
+  }
 
+  private async sendProteusMessage<T extends OtrMessage>(
+    params: SendProteusMessageParams<T>,
+    genericMessage: GenericMessage,
+    content: T['content'],
+  ): Promise<T> {
+    const {userIds, sendAsProtobuf, conversationDomain, nativePush, targetMode, payload, onClientMismatch, onSuccess} =
+      params;
     const response = await this.sendGenericMessage(
       this.apiClient.validatedClientId,
-      payloadBundle.conversation,
+      payload.conversation,
       genericMessage,
       {
         userIds,
         sendAsProtobuf,
         conversationDomain,
         nativePush,
         targetMode,
-        onClientMismatch: callbacks?.onClientMismatch,
+        onClientMismatch,
       },
     );
 
     if (!response.errored) {
       if (!this.isClearFromMismatch(response)) {
         // We warn the consumer that there is a mismatch that did not prevent message sending
-        await callbacks?.onClientMismatch?.(response, true);
+        await onClientMismatch?.(response, true);
       }
-      callbacks?.onSuccess?.(genericMessage, response.time);
+      onSuccess?.(genericMessage, response.time);
     }
 
     return {
-      ...payloadBundle,
-      content: processedContent || payloadBundle.content,
+      ...payload,
+      content,
       messageTimer: genericMessage.ephemeral?.expireAfterMillis || 0,
       state: response.errored ? PayloadBundleState.CANCELLED : PayloadBundleState.OUTGOING_SENT,
     };
   }
 
+  /**
+   * Sends a message to a conversation
+   *
+   * @param params.payload The message to send to the conversation
+   * @param params.protocol The protocol to use to send the message (MLS or Proteus)
+   * @param params.groupId? The groupId of the conversation to send the message to (Needed only for MLS)
+   * @param params.userIds? Can be either a QualifiedId[], string[], UserClients or QualfiedUserClients. The type has some effect on the behavior of the method. (Needed only for Proteus)
+   *    When given a QualifiedId[] or string[] the method will fetch the freshest list of devices for those users (since they are not given by the consumer). As a consequence no ClientMismatch error will trigger and we will ignore missing clients when sending
+   *    When given a QualifiedUserClients or UserClients the method will only send to the clients listed in the userIds. This could lead to ClientMismatch (since the given list of devices might not be the freshest one and new clients could have been created)
+   *    When given a QualifiedId[] or QualifiedUserClients the method will send the message through the federated API endpoint
+   *    When given a string[] or UserClients the method will send the message through the old API endpoint
+   * @return resolves with the sent message
```

Comment:
This doc can be moved to the SendProteusMessageParams. This will allow VScode to show the doc as you type ;)

Reply:
https://github.com/wireapp/wire-web-packages/pull/4314/commits/1eca14a201ce41c0fcaab10131fe4cf3a593aa70